dist/gcc-compatible/Hacl_Hash_Blake2b.c

```diff
@@ -679,46 +679,138 @@ void Hacl_Hash_Blake2b_finish(uint32_t nn, uint8_t *output, uint64_t *hash)
   Lib_Memzero0_memzero(b, 64U, uint8_t, void *);
 }
 
-/**
-  State allocation function when there is no key
-*/
-Hacl_Hash_Blake2b_state_t *Hacl_Hash_Blake2b_malloc(void)
+uint8_t *FStar_Pervasives_Native_snd__uint32_t__uint8_t_(K___uint32_t__uint8_t_ x)
+{
+  return x.snd;
+}
+
+static Hacl_Hash_Blake2b_state_t *malloc_raw(uint32_t kk, K___uint32_t__uint8_t_ key)
 {
   uint8_t *buf = (uint8_t *)KRML_HOST_CALLOC(128U, sizeof (uint8_t));
   uint64_t *wv = (uint64_t *)KRML_HOST_CALLOC(16U, sizeof (uint64_t));
   uint64_t *b = (uint64_t *)KRML_HOST_CALLOC(16U, sizeof (uint64_t));
-  Hacl_Hash_Blake2b_block_state_t block_state = { .fst = wv, .snd = b };
+  Hacl_Hash_Blake2b_block_state_t block_state = { .fst = kk, .snd = { .fst = wv, .snd = b } };
+  uint32_t ite;
+  if (kk != 0U)
+  {
+    ite = 128U;
+  }
+  else
+  {
+    ite = 0U;
+  }
   Hacl_Hash_Blake2b_state_t
-  s = { .block_state = block_state, .buf = buf, .total_len = (uint64_t)0U };
+  s = { .block_state = block_state, .buf = buf, .total_len = (uint64_t)ite };
   Hacl_Hash_Blake2b_state_t
   *p = (Hacl_Hash_Blake2b_state_t *)KRML_HOST_MALLOC(sizeof (Hacl_Hash_Blake2b_state_t));
   p[0U] = s;
-  Hacl_Hash_Blake2b_init(block_state.snd, 0U, 64U);
+  uint32_t kk1 = key.fst;
+  uint8_t *key1 = key.snd;
+  if (!(kk1 == 0U))
+  {
+    uint8_t *sub_b = buf + kk1;
+    memset(sub_b, 0U, (128U - kk1) * sizeof (uint8_t));
+    memcpy(buf,
+      FStar_Pervasives_Native_snd__uint32_t__uint8_t_((
+          (K___uint32_t__uint8_t_){ .fst = kk1, .snd = key1 }
+        )),
+      kk1 * sizeof (uint8_t));
+  }
+  Hacl_Hash_Blake2b_init(block_state.snd.snd, kk1, 64U);
   return p;
 }
 
 /**
-  Re-initialization function when there is no key
+  State allocation function when there is a key
 */
-void Hacl_Hash_Blake2b_reset(Hacl_Hash_Blake2b_state_t *state)
+Hacl_Hash_Blake2b_state_t *Hacl_Hash_Blake2b_malloc_with_key(uint8_t *k, uint32_t kk)
+{
+  return malloc_raw(kk, ((K___uint32_t__uint8_t_){ .fst = kk, .snd = k }));
+}
+
+/**
+  State allocation function when there is no key
+*/
+Hacl_Hash_Blake2b_state_t *Hacl_Hash_Blake2b_malloc(void)
+{
+  return Hacl_Hash_Blake2b_malloc_with_key(NULL, 0U);
+}
+
+static uint32_t fst__uint32_t__uint64_t_____uint64_t_(Hacl_Hash_Blake2b_block_state_t x)
+{
+  return x.fst;
+}
+
+static void reset_raw(Hacl_Hash_Blake2b_state_t *state, K___uint32_t__uint8_t_ key)
 {
   Hacl_Hash_Blake2b_state_t scrut = *state;
   uint8_t *buf = scrut.buf;
   Hacl_Hash_Blake2b_block_state_t block_state = scrut.block_state;
-  Hacl_Hash_Blake2b_init(block_state.snd, 0U, 64U);
+  uint32_t i = fst__uint32_t__uint64_t_____uint64_t_(block_state);
+  KRML_MAYBE_UNUSED_VAR(i);
+  uint32_t kk1 = key.fst;
+  uint8_t *key1 = key.snd;
+  if (!(kk1 == 0U))
+  {
+    uint8_t *sub_b = buf + kk1;
+    memset(sub_b, 0U, (128U - kk1) * sizeof (uint8_t));
+    memcpy(buf,
+      FStar_Pervasives_Native_snd__uint32_t__uint8_t_((
```

Comment:
Nit: Could this be inlined ?

Reply:
yeah if it were a dedicated struct with projectors marked inline_for_extraction -- this struct is going to gain more fields in the upcoming PR so I can try to remove that in the final PR

Reply:
this would require adding an alternate definition of fst and snd for HACL* that is marked inline_for_extraction... is it worth the trouble? perhaps

Reply:
or we submit an upstream PR on F*